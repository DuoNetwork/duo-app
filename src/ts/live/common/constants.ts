export * from '../../../../../duo-admin/src/constants';

export const AC_CTD_STATES = 'custodianStates';
export const AC_CTD_PRICES = 'custodianPrices';
export const AC_ACCOUNT = 'account';
export const AC_BALANCES = 'balances';
export const AC_ADDRESSES = 'addresses';
export const AC_DNM_STATUS = 'dynamoStatus';
export const AC_DMN_HOURLY = 'dynamoHourly';
export const AC_DMN_MINUTELY = 'dynamoMinutely';
export const AC_DMN_PRICES = 'dynamoPrices';
export const AC_REFRESH = 'refresh';
