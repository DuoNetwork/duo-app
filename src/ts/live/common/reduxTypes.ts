import { IBalances, ICustodianPrice, ICustodianStates } from './types';

export interface IState {
	contract: IContractState;
}

export interface IContractState {
	custodianStates: ICustodianStates;
	custodianPrices: {
		reset: ICustodianPrice;
		last: ICustodianPrice;
		// first: ICustodianPrice;
		// second: ICustodianPrice;
	};
	balances: IBalances;
}

export type Action = IBaseAction | IBooleanAction | IStringAction | IObjectAction | INumberAction;

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;

export interface IBaseAction {
	type: string;
}

export interface IBooleanAction {
	type: string;
	value: boolean;
}

export interface IStringAction {
	type: string;
	value: string;
}

export interface IObjectAction {
	type: string;
	value: object;
}

export interface INumberAction {
	type: string;
	value: number;
}

export type ThunkAction = (dispatch: Dispatch) => any;
export type PromiseAction = Promise<Action>;
