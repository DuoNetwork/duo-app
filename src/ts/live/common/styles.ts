export const ColorStyles = {
	//Background Colors
	HeaderBackground: 'rgba(32,31,33,.5)',
	CardBackground: 'rgba(32,31,33,.5)',
	CardBackgroundSolid: 'rgba(30,31,32,1)',
	HoverBackgroundSolid: 'rgba(40,41,42,1)',
	SelectBackgroundSolid: 'rgba(45,46,48,1)',
	ListHighlight: 'rgba(255,255,255,.04)',
	//Border Colors
	BorderWhite1: 'rgba(255,255,255,.1)',
	BorderWhite2: 'rgba(255,255,255,.2)',
	BorderWhite3: 'rgba(255,255,255,.3)',
	BorderWhite4: 'rgba(255,255,255,.4)',
	BorderWhite5: 'rgba(255,255,255,.5)',
	BorderWhite6: 'rgba(255,255,255,.6)',
	BorderWhite7: 'rgba(255,255,255,.7)',
	BorderWhite8: 'rgba(255,255,255,.8)',
	BorderWhite9: 'rgba(255,255,255,.9)',
	BorderWhite10: 'rgba(255,255,255,1)',
	//Text Colors
	TextWhite: 'rgba(255,255,255,1)',
	TextWhiteAlpha: 'rgba(255,255,255,.8)',
	TextWhiteAlphaL: 'rgba(255,255,255,.6)',
	TextWhiteAlphaLL: 'rgba(255,255,255,.4)',
	TextWhiteAlphaLLL: 'rgba(255,255,255,.2)',
	TextRed: 'rgba(255,94,94,1)',
	TextRedAlpha: 'rgba(255,94,94,.8)',
	TextRedAlphaL: 'rgba(255,94,94,.6)',
	TextRedAlphaLL: 'rgba(255,94,94,.4)',
	TextGreen: 'rgba(78,255,78,1)',
	TextGreenAlpha: 'rgba(78,255,78,.8)',
	TextGreenAlphaL: 'rgba(78,255,78,.6)',
	TextGreenAlphaLL: 'rgba(78,255,78,.4)',
	//Button Colors
	ButtonRadioUnchekedBG: 'rgba(55,54,56,.2)',
	ButtonRadioChekedBG: 'rgba(55,54,56,.6)',
	ButtonHoverWhite10: 'rgba(255,255,255,1)',
	ButtonHoverWhite9: 'rgba(255,255,255,0.9)',
	ButtonHoverWhite5: 'rgba(255,255,255,0.5)',
	ButtonHoverWhite4: 'rgba(255,255,255,0.4)',
	ButtonHoverWhite3: 'rgba(255,255,255,0.3)',
	ButtonHoverWhite1: 'rgba(255,255,255,0.1)',
};
