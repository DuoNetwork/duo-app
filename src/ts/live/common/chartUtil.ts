import moment from 'moment';
import * as CST from '../common/constants';
import { IAcceptedPrice, ICustodianPrice, ICustodianStates, IPriceBar, ISourceData } from './types';

class ChartUtil {
	public interpolate(sourceData: IPriceBar[], isHourly: boolean): IPriceBar[] {
		if (!sourceData.length) return [];
		sourceData.sort((a, b) => a.timestamp - b.timestamp);
		const newSourceData: IPriceBar[] = [];
		const timeStep = isHourly ? 3.6e6 : 60000;
		const source = sourceData[0].source;
		for (let i = 0; i < sourceData.length - 1; i++) {
			newSourceData.push(sourceData[i]);
			const timeIndex = sourceData[i].timestamp;
			const nextTimeIndex = sourceData[i + 1].timestamp;
			const close = sourceData[i].close;
			let newTimeStamp = timeIndex + timeStep;
			while (newTimeStamp < nextTimeIndex) {
				newSourceData.push({
					source: source,
					date: moment(newTimeStamp).format('YYYY-MM-DD'),
					hour: moment(newTimeStamp).format('HH'),
					minute: isHourly ? 0 : Number(moment(newTimeStamp).format('mm')),
					open: close,
					high: close,
					low: close,
					close: close,
					volume: 0,
					timestamp: newTimeStamp
				});
				newTimeStamp += timeStep;
			}
		}
		newSourceData.push(sourceData[sourceData.length - 1]);

		return newSourceData;
	}

	public reset(
		price: IAcceptedPrice[],
		limitUp: number,
		limitDown: number,
		limitPeriod: number
	): IAcceptedPrice[] {
		return price
			.filter(p => p.navB >= limitUp || p.navB <= limitDown || p.navA >= limitPeriod)
			.map(p => ({
				price: p.price,
				navA: 1,
				navB: 1,
				timestamp: p.timestamp + 717
			}));
	}

	public mergeReset(prices: IAcceptedPrice[], resets: IAcceptedPrice[]): IAcceptedPrice[] {
		const all = [...prices, ...resets];
		all.sort((a, b) => a.timestamp - b.timestamp);
		return all;
	}

	public mergeLastToPriceBar(
		pricebars: ISourceData<IPriceBar[]>,
		last: ISourceData<ICustodianPrice>,
		isHourly: boolean
	) {
		CST.EXCHANGES.forEach(src => {
			const pricebar: IPriceBar[] = pricebars[src.toLowerCase()];
			if (!pricebar || !pricebar.length) return;
			const sourceLast: ICustodianPrice = last[src.toLowerCase()];
			const lastBar = pricebar[pricebar.length - 1];
			const dateObj = moment.utc(sourceLast.timestamp);
			if (isHourly && lastBar.date + ' ' + lastBar.hour !== dateObj.format('YYYY-MM-DD HH'))
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: 0,
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else if (
				!isHourly &&
				lastBar.date + ' ' + lastBar.hour + ' ' + lastBar.minute !==
					dateObj.format('YYYY-MM-DD HH m')
			)
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: Number(dateObj.format('mm')),
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else {
				lastBar.high = Math.max(lastBar.high, sourceLast.price);
				lastBar.low = Math.min(lastBar.low, sourceLast.price);
				lastBar.close = sourceLast.price;
				lastBar.timestamp = sourceLast.timestamp;
			}
		});
	}

	public mergeLastToPrice(
		prices: IAcceptedPrice[],
		states: ICustodianStates,
		last: ICustodianPrice
	) {
		if (!prices.length) return;
		const lastPrice = prices[prices.length - 1];
		if (lastPrice.timestamp < last.timestamp)
			prices.push({
				price: last.price,
				navA: states.navA,
				navB: states.navB,
				timestamp: last.timestamp
			});
	}
}

const chartUtil = new ChartUtil();
export default chartUtil;
