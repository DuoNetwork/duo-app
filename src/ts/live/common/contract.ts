import infura from '../../../../../duo-admin/src/keys/infura.json';
import ContractUtil from '../../../../../duo-contract-util/src/contractUtil';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
const contract = new ContractUtil(window, '', provider, !__KOVAN__);
export default contract;
