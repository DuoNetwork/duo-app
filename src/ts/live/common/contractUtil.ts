import Web3 from 'web3';
import { Contract } from 'web3/types';
import infura from '../../../../../duo-admin/src/keys/infura.json';
import custodianAbi from '../../../../../duo-admin/src/static/Custodian.json';
import duoAbi from '../../../../../duo-admin/src/static/DUO.json';
import { IAddresses, IBalances, ICustodianPrices, ICustodianStates } from '../common/types';
import * as CST from './constants';
//import util from './util';

class ContractUtil {
	private web3: Web3;
	private duo: Contract;
	private custodian: Contract;
	private isReadOnly: boolean;
	public readonly custodianAddr: string;
	private readonly duoContractAddr: string;

	constructor() {
		if (typeof (window as any).web3 !== 'undefined') {
			this.isReadOnly = false;
			this.web3 = new Web3((window as any).web3.currentProvider);
		} else {
			this.web3 = new Web3(
				new Web3.providers.HttpProvider(
					(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) +
						'/' +
						infura.token
				)
			);
			this.isReadOnly = true;
		}
		this.custodianAddr = __KOVAN__ ? CST.CUSTODIAN_ADDR_KOVAN : CST.CUSTODIAN_ADDR_MAIN;
		this.custodian = new this.web3.eth.Contract(custodianAbi.abi, this.custodianAddr);
		this.duoContractAddr = __KOVAN__ ? CST.DUO_CONTRACT_ADDR_KOVAN : CST.DUO_CONTRACT_ADDR_MAIN;
		this.duo = new this.web3.eth.Contract(duoAbi.abi, this.duoContractAddr);
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		const store = (this.web3.currentProvider as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public convertCustodianState(rawState: string) {
		switch (rawState) {
			case CST.STATE_INCEPTION:
				return CST.CTD_INCEPTION;
			case CST.STATE_TRADING:
				return CST.CTD_TRADING;
			case CST.STATE_PRERESET:
				return CST.CTD_PRERESET;
			case CST.STATE_UP_RESET:
				return CST.CTD_UP_RESET;
			case CST.STATE_DOWN_RESET:
				return CST.CTD_DOWN_RESET;
			case CST.STATE_PERIOD_RESET:
				return CST.CTD_PERIOD_RESET;
			default:
				return CST.CTD_LOADING;
		}
	}

	public async getSystemStates(): Promise<ICustodianStates> {
		const [states, duoBalance] = await Promise.all([
			this.custodian.methods.getSystemStates().call(),
			this.getDuoBalance(this.custodianAddr)
		]);
		return {
			state: this.convertCustodianState(states[0].valueOf()),
			navA: this.fromWei(states[1]),
			navB: this.fromWei(states[2]),
			totalSupplyA: this.fromWei(states[3]),
			totalSupplyB: this.fromWei(states[4]),
			ethBalance: this.fromWei(states[5]),
			alpha: states[6].valueOf() / 10000,
			beta: this.fromWei(states[7]),
			feeAccumulated: this.fromWei(states[8]),
			periodCoupon: this.fromWei(states[9]),
			limitPeriodic: this.fromWei(states[10]),
			limitUpper: this.fromWei(states[11]),
			limitLower: this.fromWei(states[12]),
			commissionRate: states[13] / 10000,
			period: Number(states[14].valueOf()),
			iterationGasThreshold: Number(states[15].valueOf()),
			ethDuoFeeRatio: Number(states[16].valueOf()),
			preResetWaitingBlocks: Number(states[17].valueOf()),
			priceTol: Number(states[18].valueOf() / 10000),
			priceFeedTol: Number(states[19].valueOf() / 10000),
			priceFeedTimeTol: Number(states[20].valueOf()),
			priceUpdateCoolDown: Number(states[21].valueOf()),
			numOfPrices: Number(states[22].valueOf()),
			nextResetAddrIndex: Number(states[23].valueOf()),
			lastAdminTime: Number(states[24].valueOf()),
			adminCoolDown: Number(states[25]),
			usersLength: Number(states[26].valueOf()),
			addrPoolLength: Number(states[27].valueOf()),
			duoBalance: duoBalance
		};
	}

	public async getSystemAddresses(): Promise<IAddresses> {
		const addr: string[] = await this.custodian.methods.getSystemAddresses().call();
		const balances = await Promise.all(addr.map(a => this.getEthBalance(a)));
		return {
			operator: {
				address: addr[0],
				balance: balances[0]
			},
			feeCollector: {
				address: addr[1],
				balance: balances[1]
			},
			priceFeed1: {
				address: addr[2],
				balance: balances[2]
			},
			priceFeed2: {
				address: addr[3],
				balance: balances[3]
			},
			priceFeed3: {
				address: addr[4],
				balance: balances[4]
			},
			poolManager: {
				address: addr[5],
				balance: balances[5]
			}
		};
	}

	public async getSystemPrices(): Promise<ICustodianPrices> {
		const prices = await this.custodian.methods.getSystemPrices().call();
		const custodianPrices = [0, 1, 2, 3].map(i => ({
			address: prices[i * 3].valueOf(),
			price: this.fromWei(prices[1 + i * 3]),
			timestamp: prices[2 + i * 3].valueOf() * 1000
		}));

		return {
			first: custodianPrices[0],
			second: custodianPrices[1],
			reset: custodianPrices[2],
			last: custodianPrices[3]
		};
	}

	public async getBalances(address: string): Promise<IBalances> {
		if (!address)
			return {
				eth: 0,
				duo: 0,
				allowance: 0,
				tokenA: 0,
				tokenB: 0
			};

		const balances = await Promise.all([
			this.getEthBalance(address),
			this.getDuoBalance(address),
			this.getDuoAllowance(address),
			this.getTokenBalance(address, true),
			this.getTokenBalance(address, false)
		]);

		return {
			eth: balances[0],
			duo: balances[1],
			allowance: balances[2],
			tokenA: balances[3],
			tokenB: balances[4]
		};
	}

	public getUserAddress(index: number) {
		return this.custodian.methods.users(index).call();
	}

	public getPoolAddress(index: number) {
		return this.custodian.methods.addrPool(index).call();
	}

	public async getGasPrice(): Promise<number> {
		return await this.web3.eth.getGasPrice();
	}

	public async getCurrentAddress(): Promise<string> {
		return (await this.web3.eth.getAccounts())[0];
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3.eth.net.getId();
	}

	public async getEthBalance(address: string): Promise<number> {
		return this.fromWei(await this.web3.eth.getBalance(address));
	}

	private async getDuoBalance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.balanceOf(address).call());
	}

	private async getDuoAllowance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.allowance(address, this.custodianAddr).call());
	}

	private async getTokenBalance(address: string, isA: boolean): Promise<number> {
		return this.fromWei(await this.custodian.methods.balanceOf(isA ? 0 : 1, address).call());
	}

	public fromWei(value: string | number) {
		return Number(this.web3.utils.fromWei(value, 'ether'));
	}

	public toWei(value: string | number) {
		return this.web3.utils.toWei(value + '', 'ether');
	}

	public checkAddress(addr: string) {
		if (!addr.startsWith('0x') || addr.length !== 42) return false;
		return this.web3.utils.checkAddressChecksum(addr);
	}

	public create(
		address: string,
		value: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.create(payFeeInEth)
			.send({
				from: address,
				value: this.toWei(value)
			})
			.on('transactionHash', onTxHash);
	}

	public redeem(
		address: string,
		amtA: number,
		amtB: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.redeem(this.toWei(amtA), this.toWei(amtB), payFeeInEth)
			.send({
				from: address
			})
			.on('transactionHash', onTxHash);
	}

	public duoApprove(address: string, spender: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.approve(spender, this.toWei(value)).send({
			from: address
		});
	}

	public duoTransfer(address: string, to: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.transfer(to, this.toWei(value)).send({
			from: address
		});
	}

	public approve(address: string, spender: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods.approve(isA ? 0 : 1, spender, this.toWei(value)).send({
			from: address
		});
	}

	public transfer(address: string, to: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		// dummy from address
		return this.custodian.methods.transfer(isA ? 0 : 1, address, to, this.toWei(value)).send({
			from: address
		});
	}

	public collectFee(address: string, amount: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.collectFee(this.toWei(amount)).send({
			from: address
		});
	}

	public setValue(address: string, index: number, newValue: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.setValue(index, newValue).send({
			from: address
		});
	}

	public addAddress(address: string, addr1: string, addr2: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr1, addr2).send({
			from: address
		});
	}

	public removeAddress(address: string, addr: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr).send({
			from: address
		});
	}

	public updateAddress(address: string, currentRole: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(currentRole).send({
			from: address
		});
	}
}

const contractUtil = new ContractUtil();
export default contractUtil;
