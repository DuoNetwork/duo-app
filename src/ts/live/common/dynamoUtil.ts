import AWS from 'aws-sdk';
import { ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => {
				if (err) reject(err);
				else resolve(data);
			});
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output =  status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else if (process.startsWith('FEED'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_HISTORY_PRICE].N),
					volume: Number(d[CST.DB_HISTORY_VOLUME].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
