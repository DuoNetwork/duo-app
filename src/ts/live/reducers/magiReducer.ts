import { AnyAction } from 'redux';
import * as CST from '../common/constants';
import { IMagiState } from '../common/types';

export const initialState: IMagiState = {
	acceptedPrices: [],
	subscription: 0
};

export function magiReducer(
	state: IMagiState = initialState,
	action: AnyAction
): IMagiState {
	switch (action.type) {
		case CST.AC_MAG_ACCEPTED_PX:
			return Object.assign({}, state, {
				acceptedPrices: action.value
			});
		case CST.AC_MAG_SUB:
			if (action.value)
				return Object.assign({}, state, {
					subscription: action.value
				});
			else {
				window.clearInterval(state.subscription);
				return initialState;
			}
		default:
			return state;
	}
}
