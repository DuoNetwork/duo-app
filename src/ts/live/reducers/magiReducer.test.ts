import * as CST from '../common/constants';
import { initialState, magiReducer } from './magiReducer';

describe('magi reducer', () => {
	let state = initialState;

	test('default', () => {
		state = magiReducer(state, { type: 'any' });
		expect(state).toMatchSnapshot();
	});

	test('magiAcceptedPrices', () => {
		state = magiReducer(state, {
			type: CST.AC_MAG_ACCEPTED_PX,
			value: ['test']
		});
		expect(state).toMatchSnapshot();
	});

	test('magiSubscription on', () => {
		state = magiReducer(state, {
			type: CST.AC_MAG_SUB,
			value: 123
		});
		expect(state).toMatchSnapshot();
	});

	test('magiSubscription off', () => {
		window.clearInterval = jest.fn();
		state = magiReducer(state, {
			type: CST.AC_MAG_SUB,
			value: 0
		});
		expect(state).toMatchSnapshot();
		expect((window.clearInterval as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	});
});
