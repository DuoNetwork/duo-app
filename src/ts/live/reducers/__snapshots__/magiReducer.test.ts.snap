// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`magi reducer default 1`] = `
Object {
  "acceptedPrices": Array [],
  "subscription": 0,
}
`;

exports[`magi reducer magiAcceptedPrices 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "subscription": 0,
}
`;

exports[`magi reducer magiSubscription off 1`] = `
Object {
  "acceptedPrices": Array [],
  "subscription": 0,
}
`;

exports[`magi reducer magiSubscription off 2`] = `
Array [
  Array [
    123,
  ],
]
`;

exports[`magi reducer magiSubscription on 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "subscription": 123,
}
`;
